osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs

```diff
@@ -164,24 +166,42 @@ static IEnumerable<Mod> createDifficultyAdjustmentModCombinations(ReadOnlyMemory
         /// <param name="skills">The skills which processed the beatmap.</param>
-        /// <param name="clockRate">The rate at which the gameplay clock is run at.</param>
-        protected abstract DifficultyAttributes CreateDifficultyAttributes(IBeatmap beatmap, Mod[] mods, Skill[] skills, double clockRate);
+        /// <param name="clockTimeAt">The rate at which the gameplay clock is run at.</param>
+        protected abstract DifficultyAttributes CreateDifficultyAttributes(IBeatmap beatmap, Mod[] mods, Skill[] skills, Func<double, double> clockTimeAt);
 
         /// <summary>
         /// Enumerates <see cref="DifficultyHitObject"/>s to be processed from <see cref="HitObject"/>s in the <see cref="IBeatmap"/>.
         /// </summary>
         /// <param name="beatmap">The <see cref="IBeatmap"/> providing the <see cref="HitObject"/>s to enumerate.</param>
-        /// <param name="clockRate">The rate at which the gameplay clock is run at.</param>
+        /// <param name="clockTimeAt">The rate at which the gameplay clock is run at.</param>
         /// <returns>The enumerated <see cref="DifficultyHitObject"/>s.</returns>
-        protected abstract IEnumerable<DifficultyHitObject> CreateDifficultyHitObjects(IBeatmap beatmap, double clockRate);
+        protected abstract IEnumerable<DifficultyHitObject> CreateDifficultyHitObjects(IBeatmap beatmap, Func<double, double> clockTimeAt);
 
         /// <summary>
         /// Creates the <see cref="Skill"/>s to calculate the difficulty of an <see cref="IBeatmap"/>.
         /// </summary>
         /// <param name="beatmap">The <see cref="IBeatmap"/> whose difficulty will be calculated.</param>
         /// <param name="mods">Mods to calculate difficulty with.</param>
-        /// <param name="clockRate">Clockrate to calculate difficulty with.</param>
+        /// <param name="clockTimeAt">Clockrate to calculate difficulty with.</param>
         /// <returns>The <see cref="Skill"/>s.</returns>
-        protected abstract Skill[] CreateSkills(IBeatmap beatmap, Mod[] mods, double clockRate);
+        protected abstract Skill[] CreateSkills(IBeatmap beatmap, Mod[] mods, Func<double, double> clockTimeAt);
+
+        /// <summary>
+        /// Creates a function used to calculate clocktime at a point in time.
+        /// </summary>
+        /// <param name="mods">Mods to calculate clocktime with.</param>
+        /// <returns></returns>
+        protected virtual Func<double, double> BakeClockRate(Mod[] mods)
+        {
+            if (!mods.OfType<IApplicableToRate>().Any())
+                return t => t;
+
+            if(mods.OfType<IApplicableToRate>().FirstOrDefault() is ModTimeRamp rateMod)
+            {
+                return t => rateMod.GetTimeAt(t);
+            }
+
+            return t => t / mods.OfType<IApplicableToRate>().Aggregate(1.0, (prod, rateMod) => prod * rateMod.ApplyToRate(t));
```

Comment:
As above, this is not really accurate if all `IApplicableToRate` mods aren't subclasses of `ModRateAdjust` (i.e. if their rates of playback are not constant).